### ISSUE TITLE
Move to a Trie to match tlds

### ISSUE THREAD

**Issue by Yomguithereal** [barseghyanartur/tld#30]
Hello @barseghyanartur.

It is possible to enhance the performance and maybe memory footprint of your library (although I am not completely sure since needing more objects might be more RAM costly than just having a ~10k items hashset) by storing the tlds in a Trie in reverse order. This would mean more straightforward queries (just need to query the Trie to find the tld + more efficient misses if the query is not a tld etc.). I am not completely sure it would be more efficient though because it may mean more set/dict lookups than the current version but it depends on the average number of stems in the domain string.

What do you think? I can draft a PR to implement this if you feel this is a good way to go.

**Comment by Yomguithereal** [comment]
@barseghyanartur I want to fix the other issues listed on the repo to help you. Should I do that on my trie branch or do you feel it would be better to make them on the non-trie one for the time being?

**Comment by barseghyanartur** [comment]
@Yomguithereal:

I think it's better to separate. The other issues are likely to land in the nearest releases earlier. What issues would you like to work on? I think at the moment, #32 (and related #27) are the most important.

#32 we have already [discussed](https://github.com/barseghyanartur/tld/issues/32#issuecomment-391295841).
#27 is pretty clear too; see my [comment](https://github.com/barseghyanartur/tld/issues/27#issuecomment-391835286).

**Comment by Yomguithereal** [comment]
Hello @barseghyanartur. I did a quick benchmark to see if the Trie was any good. I processed csv files containing 1,5M & 3M respectively using this code:

```python
import csv
from src.tld import get_tld

with open('./urls.csv', 'r') as f:
    reader = csv.reader(f)
    next(reader)

    for line in reader:
        for url in line[1].split('|'):
            t = get_tld(url, fail_silently=True)
```

Here are the results:

```
1,5M
TRIE:   python bench.py  207,89s user 1,09s system 99% cpu 3:29,18 total
MASTER: python bench.py  246,39s user 1,12s system 99% cpu 4:07,72 total

3M
TRIE:   python bench.py  425,43s user 2,41s system 99% cpu 7:08,36 total
MASTER: python bench.py  508,62s user 3,65s system 99% cpu 8:33,20 total
```

So nothing overwhelming but the Trie does have an edge.

**Comment by Yomguithereal** [comment]
My problem here is that I feel I will have to code things twice if I start fixing the other issues.

**Comment by barseghyanartur** [comment]
@Yomguithereal:

I understand. Fair enough. 

My proposal:

- Could you push benchmarks as well (so that I also can test it)?
- After checking benchmarks, I will create a new branch for the 0.8 version in which I will merge the Trie changes. That will also be the branch for all other 0.8 improvements.

Agree? :)

**Comment by Yomguithereal** [comment]
The only problem with this is that I can't push my csv file. It's way too big. I will first design a less real-case scenario benchmark instead on loops etc.

**Comment by barseghyanartur** [comment]
What about pushing it to another repository and downloading once when testing into temp directory?

**Comment by Yomguithereal** [comment]
1) they are gigabytes
2) they are from twitter data and I don’t know if I can freely open them (probably not)

**Comment by barseghyanartur** [comment]
@Yomguithereal:

In that case I suggest to use [Faker](https://github.com/joke2k/faker) in order to generate domain names.

**Comment by Yomguithereal** [comment]
Seems like a good idea. I'll try to make something quick.

**Comment by barseghyanartur** [comment]
@Yomguithereal:

Branch [v0.8](https://github.com/barseghyanartur/tld/tree/v0.8) shall be used for work-in-progress. It does contain the Trie changes and changes for #27 and #32 will be merged there (for testing).

**Comment by barseghyanartur** [comment]
Released in 0.8.
